### PR TITLE
Fix for #2280 - Cannot do an "add" patch to IList<Anything>

### PR DESCRIPTION
**Pull request by kirthik**
@rynowak - could you please help take a look?

### CODE REVIEW

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 using Microsoft.AspNet.JsonPatch.Operations;
 using Newtonsoft.Json;
 using Newtonsoft.Json.Serialization;
+using System.Collections.Generic;
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: sort

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs [hunk 2]
```diff
@@ -638,9 +642,22 @@ public void Copy(Operation<T> operation, T objectToApplyTo)
...
+            if (jsonPropertyType.IsGenericType)
+            {
+                var genericArguments = jsonPropertyType.GetGenericArguments();
+                if (genericArguments.Length != 1)
+                {
+                    return false;
+                }
+                var listType = typeof(IList<>).MakeGenericType(genericArguments);
+                return listType.GetTypeInfo().IsAssignableFrom(jsonPropertyType);
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: The right check here is probably to look at `.ImplementedInterfaces` for a type which is is generic and has `IList<>` as it's generic definition. This code here isn't really accurate.

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs [hunk 3]
```diff
@@ -133,9 +134,9 @@ private void Add(string path, object value, T objectToApplyTo, Operation<T> oper
                 // what if it's an array but there's no position??
                 if (IsNonStringArray(patchProperty))
                 {
-                    // now, get the generic type of the enumerable
+                    // now, get the generic type of the enumerable from the runtime type of patchProperty value.
                     var genericTypeOfArray = PropertyHelpers.GetEnumerableType(
-                        patchProperty.Property.PropertyType);
+                        patchProperty.Property.ValueProvider.GetValue(patchProperty.Parent).GetType());
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: What specific case is this needed for? What if the value is null.

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs [hunk 4]
```diff
@@ -133,9 +134,9 @@ private void Add(string path, object value, T objectToApplyTo, Operation<T> oper
                 // what if it's an array but there's no position??
                 if (IsNonStringArray(patchProperty))
                 {
-                    // now, get the generic type of the enumerable
+                    // now, get the generic type of the enumerable from the runtime type of patchProperty value.
                     var genericTypeOfArray = PropertyHelpers.GetEnumerableType(
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: Should we be looking for an `IList<>` type instead? That's really the contract that we need if we're adding/inserting.

File: test/Microsoft.AspNet.JsonPatch.Test/NestedObjectTests.cs [hunk 5]
```diff
@@ -202,6 +202,151 @@ public void AddToListWithSerialization()
         }
 
         [Fact]
+        public void AddToIntListToGeneralList()
```
Comment (rynowak) [inline test/Microsoft.AspNet.JsonPatch.Test/NestedObjectTests.cs]: Do all of these tests vary on the declared type of the property? (Didn't you update the code to use the runtime type).

I'd also like to see one where we're modifying a list through a nested property.

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs (hunk 3 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: `JsonProperty`'s `PropertyType` returns declared type of the property not the runtime type. So if there is a property of declared type `IList`, this call returns `null`. So I had to change it to take the  runtime type of value.

Its a good point, I have to handle the case where runtime type value is `null`.

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs (hunk 4 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: If Property's declared type is `IList<>`, then `.ImplementedInterfaces` would not have `IList<>`. Getting generic type from `IEnumerable` works when Property is of type `'IList` or `IList<>` or `List<>`.

File: test/Microsoft.AspNet.JsonPatch.Test/NestedObjectTests.cs (hunk 5 above)
Comment (kirthik) [inline test/Microsoft.AspNet.JsonPatch.Test/NestedObjectTests.cs]: `JsonProperty`'s `PropertyType` is fetching declared type of the property not the runtime type.

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs (hunk 1 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: Forgot to sort. Will do it in the next iteration :)

**Comment by kirthik** [comment]
@rynowak  : Updated code based on comments. Could you please take a look?

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs [hunk 6]
```diff
@@ -636,11 +637,14 @@ public void Copy(Operation<T> operation, T objectToApplyTo)
             }
         }
 
-        private bool IsNonStringArray(JsonPatchProperty patchProperty)
+        private bool IsNonStringArray(Type type)
         {
-            return !(patchProperty.Property.PropertyType == typeof(string))
-                    && typeof(IList).GetTypeInfo().IsAssignableFrom(
-                        patchProperty.Property.PropertyType.GetTypeInfo());
+            if (PropertyHelpers.GetIListType(type) != null)
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: If this is only used here please just inline the code. We don't want to litter our code with random helpers when a private method does what you want already.

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs (hunk 6 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: `GetIListType` is used in multiple places in this class. I just updated this code to re-use existing method.

Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: Is it used in multiple classes?

Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs]: No. Just in this class.

**Comment by kirthik** [comment]
Checked in - 3bc7599903dfd8ddc55d9249794efa65b2f527e1

**Pull request closed without merging**